power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/dataset.hpp

```diff
@@ -268,6 +428,29 @@ template <dataset_type_tag dataset_type_> class Dataset {
         return std::span<StructType>{ptr + info.elements_per_scenario * scenario,
                                      ptr + info.elements_per_scenario * (scenario + 1)};
     }
+
+    // get non-empty columnar buffer
+    template <class StructType>
+    RangeObject<StructType> get_columnar_buffer_span_impl(Idx scenario, Idx component_idx) const {
+        // return empty span if the component does not exist
+        if (component_idx < 0) {
+            return {};
+        }
+        // return span based on uniform or non-uniform buffer
+        ComponentInfo const& info = dataset_info_.component_info[component_idx];
+        Buffer const& buffer = buffers_[component_idx];
+        assert(buffer.data == nullptr);
+        RangeObject<StructType> const total_range{info.total_elements, buffer.attributes};
+        if (scenario < 0) {
+            return total_range;
+        }
+        if (info.elements_per_scenario < 0) {
+            return RangeObject<StructType>{total_range.begin() + buffer.indptr[scenario],
+                                           total_range.begin() + buffer.indptr[scenario + 1]};
+        }
+        return RangeObject<StructType>{total_range.begin() + info.elements_per_scenario * scenario,
+                                       total_range.begin() + info.elements_per_scenario * (scenario + 1)};
```

Comment:
this still contains a bug. see my other comment. Maybe that's also the reason you were struggling with combining the implementation with the row-based buffer

Reply:
This implementation should be fine now. The tests are a little bit simplified as well, since taking sub-ranges now works as intended.